components/clm/src/biogeochem/LSparseMatMod.F90

```diff
@@ -0,0 +1,428 @@
+module LSparseMatMod
+
+
+  ! !USES:
+  ! sparse matrix capability
+  use bshr_kind_mod , only : r8 => shr_kind_r8
+  use bshr_log_mod  , only : errMsg => shr_log_errMsg
+  use clm_varctl    , only : iulog
+implicit none
+  private
+  type, public :: spm_list_type
+    real(r8) :: val
+    integer  :: icol
+    integer  :: irow
+    type(spm_list_type), pointer :: next => null()
+  end type spm_list_type
+
+  type :: lom_type
+  contains
+    procedure :: calc_state_pscal
+    procedure :: calc_reaction_rscal
+    procedure :: apply_reaction_rscal
+  end type lom_type
+  type, public :: sparseMat_type
+    real(r8), pointer :: val(:)      !nonzero val
+    integer , pointer :: icol(:)     !col id for each val
+    integer , pointer :: pB(:)       !first nonzero column id (as in val) in each row
+    integer , pointer :: ncol(:)     !number of nonzero columns in eacho row
+    integer :: szcol                 ! number of columns of the matrix
+    integer :: szrow
+  contains
+    procedure, public :: init
+  end type sparseMat_type
+  real(r8), parameter :: tiny_val=1.e-14_r8
+  public :: spm_list_init, spm_list_insert, spm_list_to_mat
+  public :: flux_correction
+contains
+
+!---------------------------------------------------------------
+  function create_spm_type()
+    !
+    ! ! DESCRIPTION:
+    !
+    ! create an object of type sparseMat_type.
+    ! Right now it is purposely empty
+   class(sparseMat_type), pointer :: spm
+   class(sparseMat_type), pointer ::  create_spm_type
+   allocate(spm)
+   create_spm_type => spm
+  end function create_spm_type
+
+!---------------------------------------------------------------
+  subroutine init(this,nelms,nrows,szcol)
+  !
+  !DESCRIPTION
+  ! initialize the sparse matrix
+  implicit none
+  class(sparseMat_type), intent(inout) :: this
+  integer, intent(in) :: nelms
+  integer, intent(in) :: nrows
+  integer, intent(in) :: szcol
+
+  allocate(this%val(nelms)) ;  this%val(:)  = 0._r8
+  allocate(this%icol(nelms));  this%icol(:) = 0
+  allocate(this%pB(nrows));    this%pB(:) = 0
+  allocate(this%ncol(nrows));  this%ncol(:) = 0
+  this%szcol = szcol
+  this%szrow= nrows
+  end subroutine init
+
+!---------------------------------------------------------------
+  subroutine spm_pack(Amat, nelms)
+
+  implicit none
+  real(r8), dimension(:,:),intent(in) :: Amat
+  integer, intent(in) :: nelms
+
+  class(sparseMat_type), pointer :: spm
+  integer :: nrows, szcol
+  integer :: ii,jj,kk
+  logical :: first
+  nrows=size(Amat,1)
+  szcol=size(Amat,2)
+  allocate(spm, source=create_spm_type())
+
+  call spm%init(nelms, nrows, szcol)
+
+  kk = 0
```

Comment:
Seems like a comment here, or at the top of the routine, would be useful to explain why we do this.